packages/theming-core/src/generics/themeRegistry.ts

```diff
@@ -0,0 +1,247 @@
+/**
+ * Function which takes a parent theme, does some processing on it, and returns a new theme definition.
+ * This definition will then be passed to the standard resolver
+ *
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type ProcessTheme<ITheme, IThemeDefinition> = (parentTheme: ITheme) => IThemeDefinition;
+
+/**
+ * Function which takes a theme definition and a parent theme and produces a new resolved theme
+ *
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type ResolveTheme<ITheme, IThemeDefinition> = (definition: IThemeDefinition | undefined, parentTheme: ITheme) => ITheme;
+
+/**
+ * Graph node in teh theme graph
+ */
+interface IThemeEntry<ITheme, IThemeDefinition> {
+  /** parent theme */
+  parent?: string;
+
+  /** definition or function used to regenerate the theme */
+  definition?: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>;
+
+  /** resolved theme */
+  resolved?: ITheme;
+}
+
+/** Predefined entries in the theme graph, pretty self-explanatory */
+interface IThemeGraphCore<ITheme, IThemeDefinition> {
+  _platform: IThemeEntry<ITheme, IThemeDefinition>;
+  default: IThemeEntry<ITheme, IThemeDefinition>;
+}
+
+/** Key-names for the platform and default themes */
+const _platformKey: keyof IThemeGraphCore<object, object> = '_platform';
+const _defaultKey: keyof IThemeGraphCore<object, object> = 'default';
+
+/** Graph interface, takes the predefined entries and adds the index signature */
+interface IThemeGraph<ITheme, IThemeDefinition> extends IThemeGraphCore<ITheme, IThemeDefinition> {
+  [key: string]: IThemeEntry<ITheme, IThemeDefinition>;
+}
+
+/**
+ * Function signature for theme registration
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type IRegisterTheme<ITheme, IThemeDefinition> =
+  (definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>, name?: string, parent?: string) => void;
+
+/**
+ * Function signature for get theme calls
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type IGetTheme<ITheme> = (name?: string) => ITheme;
+
+/**
+ * Function signature for updating the platform defaults
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type IUpdatePlatformDefaults<ITheme> = (platformDefaults: ITheme) => void;
```

Comment:
I'm seeing a general theme for these types of subject and action and would propose some naming changes for consistency:

| Existing| Proposed |
| ------ | -------- |
| `ProcessTheme` | `IThemeProcessor` |
| `ResolveTheme` | `IThemeResolver` |
| `IGetTheme` | `IThemeAccessor` |
| `IRegisterTheme` | `IThemeRegistrar` |
| `IUpdatePlatformDefaults` | `IPlatformDefaultsUpdater` |

I think this fits nicely with the other `ITheme` types that are structures, as well.